src/lib/raft/raft.c

```diff
@@ -695,14 +721,51 @@ raft_worker_handle_io(struct raft *raft)
 				raft_sm_wait_election_end(raft);
 			} else {
 				/* No leaders, no votes. */
-				raft_sm_schedule_new_vote(raft, raft->self);
+				raft_sm_schedule_new_vote(raft, raft->self,
+							  raft->vclock);
 			}
 		}
 	} else {
 		memset(&req, 0, sizeof(req));
 		assert(raft->volatile_term >= raft->term);
-		req.term = raft->volatile_term;
+		/*
+		 * Vote and term bumps are persisted separately for foreign
+		 * nodes. This serves as a flush of all transactions going to
+		 * WAL right now so as the local node could correctly compare
+		 * its own vclock vs candidate's one. Otherwise the local vclock
+		 * can be <= candidate's now, but that can change after the WAL
+		 * queue is flushed.
+		 */
+		if (raft->volatile_vote == raft->vote ||
```

Comment:
I guess `volatile_vote == vote` is impossible in this branch (the only case is volatile_vote == 0).

Reply:
It is possible if you vote for self + bump term in the same message. Anyway, I reworked this part.